Develop to Master merge

iwanlevin opened the pull request:
Merge develop to master

iwanlevin committed:
database interface connection and pqxx connect try catch error handling

iwanlevin committed:
database interface table creation rough draft

brianzhouzc committed:
Skeleton function for updating checksum in DB

brianzhouzc committed:
Compute md5 with Orthanc function

brianzhouzc committed:
Merge branch 'develop' into feat-phi

brianzhouzc committed:
Fix Database issue with GitHub Actions

brianzhouzc committed:
Added workflow_dispatch to allow manually running Github Actions

brianzhouzc committed:
Add feat-phi to github action branch

brianzhouzc committed:
Output for postgres

brianzhouzc committed:
docker logs for postgres

brianzhouzc committed:
Added health check for postgres

brianzhouzc committed:
Update build.yml

brianzhouzc committed:
Merge pull request #95 from BCCF-UBCO-AD/patch-github_actions_database

Fixed Database issue with GitHub Actions

brianzhouzc committed:
Updated docker-compose with health-check

brianzhouzc committed:
Fix issue with hostname

cppcooper committed:
Moves UNIT_TEST definition to common.h

cppcooper committed:
Merge branch 'develop' into feat-phi

cppcooper committed:
Adds default log levels for branches of DEBUG() macro

brianzhouzc committed:
Merge pull request #61 from BCCF-UBCO-AD/feat-phi

Adds database interface

brianzhouzc committed:
Merge branch 'develop' into feat-checksum

raouft committed:
Add week 16 scrum report

raouft committed:
Merge branch 'documents' of https://github.com/BCCF-UBCO-AD/Orthanc-TMI into documents

raouft committed:
Add small clarification

brianzhouzc committed:
Debug message for checksum table

brianzhouzc committed:
Remove POSTGRES_HOST_AUTH_METHOD: trust

`POSTGRES_HOST_AUTH_METHOD: trust` allows all connections to be established without needing the password. Removing this before I forget about it.

cppcooper committed:
Revises debug code

brianzhouzc committed:
Merge branch 'develop' into feat-checksum

cppcooper committed:
Adds missing de-initialization

cppcooper committed:
Corrects docker file

cppcooper committed:
Adds callbacks to initialization

cppcooper committed:
Adds JobQueue class

cppcooper committed:
Improves the elegance of the job queue process method

cppcooper committed:
Adds condition_variable to sleep thread

Never done this before, so this should definitely be tested

raouft committed:
Add week 17 scrum report

raouft committed:
Fix date

raouft committed:
Update progress and future tasks

brianzhouzc committed:
Debug messages for OnStoredInstanceCallback

brianzhouzc committed:
Added a second Orthanc instance acting as a modality

cppcooper commented on .github/docker-compose.yml:
```diff
@@ -15,13 +15,19 @@ services:
     networks:
       - postgres-network
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready -U postgres"]
```
what are these?

cppcooper commented on src/plugin.cpp:
```diff
@@ -20,6 +23,16 @@ extern "C" {
     int32_t OrthancPluginInitialize(OrthancPluginContext* context){
         globals::context = context;
+        /* Connect with database interface. */
+        DBInterface::connect("postgres", "example");
```
We never disconnect. >_> `void OrthancPluginFinalize(){}`

cppcooper commented on tests/common.h:
```diff
@@ -1,7 +1,9 @@
 #pragma once
 #define UNIT_TEST
+#include <core.h>
```
We should maybe refactor this name, what do you think? Something like `unit-tests.h` maybe?

cppcooper reviewed (commented):
Looks mostly ok, I think it'd be largely ok to merge into master already. We should disconnect the DB first though.

edit: so long as the changes to the docker files are totally ok as well

cppcooper committed:
Imports md5 implementation

cppcooper committed:
Fixes include for md5

cppcooper committed:
Adds job definition for JobQueue

cppcooper committed:
Revises plugin.cpp a smidgen

cppcooper committed:
Organizes source files

cppcooper committed:
Merge branch 'random-changes' into feat-datetrunc-integration

cppcooper committed:
Removes useless include

cppcooper committed:
Adds tests-jobqueue unit test, and a missing state flip in Process()

cppcooper committed:
Adds missing include

Strange that it worked fine on my end but not in CI

cppcooper committed:
Superficial changes to Date Truncation

cppcooper committed:
Moves a problematic global variable that shouldn't be global

cppcooper committed:
Replaces loop with sleep in unit test, as the segfault wasn't the sleep

brianzhouzc committed:
Merge branch 'feat-jobqueue' into feat-checksum

cppcooper committed:
Revises WriteDicomFile and DicomFile::Write

cppcooper committed:
Adds Key to Hex and Hex to Key functions (key: 0010,0030)

cppcooper committed:
Adds PluginConfigurer::GetDateFormat(tag = 0)

brianzhouzc committed:
calc_checksum working, but update_checksum still have issues.

brianzhouzc committed:
Fix missing include

brianzhouzc committed:
Merge pull request #98 from BCCF-UBCO-AD/feat-orthanc_client

Added a second Orthanc instance acting as a modality

cppcooper committed:
Fixes potential segfault

If a "spurious" wakeup were to happen the code would have gone forward as if there were jobs to perform.. then a segmentation fault would have happened when `jqueue.pop()` was executed.
Now it'll just safely repeat the wait.

raouft committed:
Add week 18 scrum report

cppcooper committed:
Prevents redefinition warning

cppcooper committed:
Improves hex unit test

cppcooper committed:
Removes useless braces

iwanlevin committed:
changed the date truncation function to use a const char* input of the format instead of a json object that has to be passed to the function

iwanlevin committed:
small fix in datetruncate function args

iwanlevin committed:
change default date truncation to YYYY0101

iwanlevin committed:
change default date truncation to YYYYMMDD and added default test case to unit test for date truncate

raouft committed:
Merge pull request #105 from BCCF-UBCO-AD/documents

Add scrum reports to develop

cppcooper committed:
Refactors function name and parameter name

cppcooper committed:
Revises DicomFile to support buffer modification and reliable Write use

cppcooper committed:
Refactors DicomElementView constructor and adds GetValueIndex()

cppcooper committed:
Replaces DicomFilter with DicomAnonymizer (integrates truncation)

cppcooper committed:
Reintegrates MakeHardlinks

cppcooper committed:
Removes dead friend

cppcooper committed:
Adds PluginConfigurer::UnitTestInitialize() to help unit tests configure

cppcooper committed:
Adds friend class for unit test to use

cppcooper committed:
Defines tag_code = uint64_t

cppcooper committed:
Adds missing edge case to DicomAnonymizer::Truncate

cppcooper committed:
Fixes placement of "common.h" in includes

cppcooper committed:
Adds more complete logic for checking date truncation

cppcooper committed:
Formatting corrections

cppcooper committed:
Removes superfluous constructors from DicomFile

cppcooper committed:
Fixes DicomFile missing size when allocating a buffer inside the object

cppcooper committed:
Adds simple safety check for TruncateDate()

cppcooper committed:
Updates date truncation format

cppcooper committed:
Updates unit test

cppcooper committed:
Replaces move assignment with swap function

cppcooper committed:
Adds debug method to DicomAnonymizer to print the black/white lists

cppcooper committed:
Disables date truncation

cppcooper committed:
Fixes PluginConfigurer initialization code

cppcooper committed:
Adds sanitizers to cmake Debug config

cppcooper committed:
Replaces UB in DicomElementView

cppcooper committed:
Adds one missing edge case

There are more that were not added.

cppcooper committed:
Revises buffer owning DicomFile to use shared_ptr instead of unique

cppcooper committed:
Revises/fixes a few things in DicomAnonymizer

cppcooper committed:
Renames/revises tests-filter => tests-anon

cppcooper committed:
Renames test-datetruncation => tests-date-truncation

cppcooper committed:
Updates refactored code

cppcooper committed:
Fixes tests-postgrsql

cppcooper committed:
Rewrites TruncateDate to simply satisfy the missing edge cases

cppcooper committed:
Completes integration of TruncateDate

cppcooper committed:
Disables load line in anon test

cppcooper committed:
Reorganizes unit tests

cppcooper committed:
Renames files for dicom testing to change the order of suite's execution

cppcooper committed:
Reverts back to bool Anonymize(&)

cppcooper committed:
Revises DicomFile

cppcooper committed:
Adds test info prints to unit tests

cppcooper committed:
Merge branch 'random-changes' into pr-merge

cppcooper committed:
Merge branch 'feat-jobqueue' into pr-merge

cppcooper committed:
Merge branch 'feat-datetrunc-integration' into pr-merge

cppcooper committed:
Fixes build

cppcooper committed:
Fixes include order for test source

cppcooper committed:
Moves jobqueue test

cppcooper committed:
Improves DicomAnonymizer thread safety

cppcooper committed:
Adds a few important dicom tag enums

cppcooper committed:
Updates json format

cppcooper committed:
Updates DicomFile::MakeHardlinks()

cppcooper committed:
Adds debug line for double processing issue

cppcooper committed:
Adds missing static declaration

cppcooper committed:
Updates todo comments

cppcooper committed:
Updates date truncation helper get_days_for_month(m,y)

cppcooper committed:
Renames macro definitions for Verbosity to DEBUG_#

cppcooper committed:
Deletes third-party/md5 code

cppcooper committed:
Adds class comment block to dicom-anonymizer.h

cppcooper committed:
Changes json format for hardlinks

cppcooper committed:
Standardizes declaration of tags (types/names)

cppcooper committed:
Refactors DicomAnonymizer to be thread safe as integrated

cppcooper committed:
Adds code to save/retrieve data used for Hardlink folder group names

cppcooper committed:
Completes DicomFile::MakeHardlinks()

cppcooper committed:
Merge pull request #106 from BCCF-UBCO-AD/pr-merge

Merged PRs

cppcooper committed:
Reverts MakeHardlinks loop to use structured bindings

I checked that the bindings can work via a simple unit test which was thrown out after verification.
It seems the json library throws an exception if the identifier is asked to convert into the wrong type.
Subsequently it seems that the type the IDE claims is false and it is actually more complicated than the structured bindings going straight to the types needed.

brianzhouzc committed:
Using pointer as key

brianzhouzc committed:
Merge branch 'develop' into feat-checksum

brianzhouzc committed:
Cleanup after merge

brianzhouzc committed:
Fix missing include

cppcooper committed:
Fixes build error in dicom-file.cpp

brianzhouzc committed:
New docker image with official Orthanc binary

cppcooper committed:
Adds environment variable to docker for asan

cppcooper committed:
Adds and updates debug logging

cppcooper committed:
Fixes bug in PluginConfigurer::Initialize

cppcooper committed:
Adds debug/release to CI. release:data-anonymizer;debug:all-tests

cppcooper committed:
Merge pull request #111 from BCCF-UBCO-AD/feat-docker

Docker updates

cppcooper committed:
Merge branch 'develop' into refactor-dicom-anonymizer

cppcooper committed:
Merge branch 'refactor-dicom-anonymizer' into feat-hardlinks

cppcooper committed:
Merge pull request #110 from BCCF-UBCO-AD/feat-hardlinks

Feat hardlinks

cppcooper committed:
Fixes memory overlap warning

brianzhouzc committed:
Merge branch 'develop' into feat-checksum

cppcooper committed:
Updates DBInterface method names and removes a couple

cppcooper committed:
Changes header includes to follow normal convention

cppcooper committed:
Updates DicomChecksum

- Adds thread safety
- Fixes method naming convention
- Integrates DBInterface

brianzhouzc committed:
Merge branch 'feat-checksum' of https://github.com/BCCF-UBCO-AD/Orthanc-TMI into feat-checksum

cppcooper committed:
Removes DicomFile::GetData replaces with md5 calculation

cppcooper committed:
Refactors DicomFile's GetData => GetElementValue

cppcooper committed:
Adds missing statement to CalculateMd5

cppcooper committed:
Cleans up includes for storage-area.cpp

cppcooper committed:
Merge pull request #109 from BCCF-UBCO-AD/refactor-dicom-anonymizer

Refactor of DicomAnonymizer + hard links

cppcooper committed:
Adds json flag to disable/enable hash bins for hardlinks

cppcooper committed:
Consolidates PluginConfigurer's two Initialize methods

cppcooper committed:
Fixes hardlink paths (missing '/' on the group portion)

cppcooper committed:
Merge branch 'develop' into feat-checksum

cppcooper committed:
Merge branch 'fixes-hardlinks' into feat-checksum

cppcooper committed:
Fixes mistake in hardlink config parsing code

brianzhouzc committed:
Debug github actions

brianzhouzc committed:
Update github actions

brianzhouzc committed:
Cache Postgres to avoid initialization

brianzhouzc committed:
Experiment with Github Action cache

brianzhouzc committed:
Fix missing use in CI

brianzhouzc committed:
Remove unnecessary volume mount for CI

brianzhouzc committed:
Remove postgres initialization

brianzhouzc committed:
Lower sleep time for CI

cppcooper committed:
Lowers sleep to 2s and combines builds

brianzhouzc committed:
Update build.yml

brianzhouzc committed:
Merge pull request #99 from BCCF-UBCO-AD/feat-checksum

Feat checksum merge to Develope

cppcooper committed:
Disables jobqueue from building

cppcooper committed:
Revises DBInterface::connect to take all variable options

ie. database,host,port,username,password

cppcooper committed:
Moves DBInterface connection into plugin-configure.cpp

cppcooper committed:
Reorganizes files

cppcooper committed:
Revises DicomChecksum in preparation for the incoming filter callback

cppcooper committed:
Adds incoming-filter.cpp

cppcooper committed:
Revises includes

cppcooper committed:
Revises DBInterface to use member variable instead of global pointer

cppcooper committed:
Adds include/plugin to include directories

cppcooper committed:
Reorganizes files, renames DicomChecksum=>DataTransport, fixes some includes

cppcooper committed:
Adds code to reject duplicate files from being processed

cppcooper committed:
Fixes incorrect json type cast/retrieval

cppcooper committed:
Updates unit test

cppcooper committed:
Replaces retrieved host with "localhost"

cppcooper committed:
Fixes weird function invocation exception, was related to not using ptr

cppcooper committed:
Fixes warnings

cppcooper committed:
Fixes includes that were missed

cppcooper committed:
Adds INFO message about duplicate DICOM file being detected and rejected

brianzhouzc committed:
Added crosswalk table and update method

brianzhouzc committed:
Fixed missing file emplace

brianzhouzc committed:
Invert duplicate check in FilterCallback

brianzhouzc committed:
Merge branch 'refactor-codebase' into feat-crosswalk

brianzhouzc committed:
Update crosswalk table using triggers instead of callback

brianzhouzc committed:
Added SQL for phi mismatch detection

cppcooper committed:
Removes extra .DCM from hardlinks and prints operation with paths to DEBUG_1

cppcooper committed:
Removes old json keys

cppcooper committed:
Moves comment

cppcooper committed:
Revises MakeHardlinks to match GetPath's use of storage_root

brianzhouzc committed:
Remove redundant debug message

brianzhouzc committed:
Merge pull request #115 from BCCF-UBCO-AD/fix-hardlinks

Fix hardlinks

brianzhouzc committed:
Merge branch 'refactor-codebase' into feat-crosswalk

brianzhouzc committed:
Cleanup a debug output

brianzhouzc committed:
Fix crosswalk table foreign key issue

brianzhouzc committed:
Merge pull request #114 from BCCF-UBCO-AD/feat-crosswalk

Feat crosswalk

brianzhouzc committed:
Jobqueue for database initialization

cppcooper committed:
Revises UpdateChecksum signature

brianzhouzc committed:
Jobqueue for updatechecksum

cppcooper committed:
Simplifies

cppcooper committed:
Updates UpdateChecksum prepared statement

cppcooper committed:
Auto-formats SQL

cppcooper committed:
Updates prepared statement usage

cppcooper committed:
Guts DBInterface

cppcooper committed:
Revises DBInterface and integration with PluginConfigurer

cppcooper committed:
Updates DBInterface usages

cppcooper committed:
Adds missing static keyword

cppcooper committed:
Revises JobQueue to always accept jobs, even if not running

cppcooper committed:
Removes return type of AddJob

cppcooper committed:
Adds back sleep for Initialize job

cppcooper committed:
Removes JobQueue again and calls abort if DB init job fails

cppcooper commented:
Reminder to update the version number. If we have all the features, 1.0 else 0.9

brianzhouzc committed:
Separate initialization sql file

brianzhouzc committed:
SQL function for disabling crosswalk update and phi mismatch detection

cppcooper committed:
Merge pull request #113 from BCCF-UBCO-AD/refactor-codebase

Refactors codebase + fixes double processing

brianzhouzc committed:
Added function for adding records to crosswalk table

brianzhouzc committed:
Update initialize.sql

brianzhouzc committed:
Remove unnecessary constraint on database

brianzhouzc committed:
Updated InsertCrosswalk function

cppcooper committed:
Integrates DBInterface::InsertCrosswalk

cppcooper committed:
Adds missing install to cmake

cppcooper committed:
Updates readme

cppcooper committed:
Updates readme

cppcooper committed:
Merge remote-tracking branch 'origin/master' into feat-crosswalk

cppcooper committed:
Updates readme

cppcooper committed:
Updates readme

cppcooper committed:
Updates readme

cppcooper committed:
Updates readme

cppcooper committed:
Merge pull request #118 from BCCF-UBCO-AD/feat-crosswalk

Feat crosswalk

brianzhouzc committed:
New initialization script

brianzhouzc committed:
Merge branch 'feat-crosswalk' of https://github.com/BCCF-UBCO-AD/Orthanc-TMI into feat-crosswalk

brianzhouzc committed:
Fix sql errors

brianzhouzc committed:
Updated sql queries to accept file uuid instead

brianzhouzc committed:
Trim fields when inserting into crosswalk

brianzhouzc committed:
Added remaining phi mismatch cases

cppcooper committed:
Merge pull request #119 from BCCF-UBCO-AD/feat-crosswalk

Feat crosswalk

cppcooper committed:
Update configuration.h

cppcooper committed:
Updates DFD-1

cppcooper committed:
Adds report

cppcooper commented on .idea/modules.xml:
```diff
@@ -2,6 +2,7 @@
 <project version="4">
   <component name="ProjectModuleManager">
     <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/Orthanc-TMI.iml" filepath="$PROJECT_DIR$/.idea/Orthanc-TMI.iml" />
       <module fileurl="file://$PROJECT_DIR$/.idea/Orthanc_TMI.iml" filepath="$PROJECT_DIR$/.idea/Orthanc_TMI.iml" />
     </modules>
   </component>
```
Not sure how any `.idea/` files got in. I thought the directory was ignored in .gitignore.

We should delete this file and any others in there.

cppcooper commented on CMakeLists.txt:
```diff
@@ -1,5 +1,9 @@
 cmake_minimum_required(VERSION 3.20)
 set(CMAKE_CXX_STANDARD 20)
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+    add_compile_options("-fsanitize=address,undefined")
+    add_link_options("-fsanitize=address,undefined")
+endif()
```
Couldn't add the thread sanitizer, apparently they conflict amongst each other.

cppcooper commented on docker/images/orthanc/Dockerfile:
```diff
@@ -4,12 +4,19 @@ MAINTAINER Brian Zhou <brianzhouzc@gmail.com>
+RUN pacman -Sy dos2unix --noconfirm
 
 ADD ./setup-environment.sh ./setup-environment.sh
+RUN dos2unix ./setup-environment.sh
```
surely there is a better solution

cppcooper commented on docker/orthanc/orthanc.json:
```diff
@@ -14,22 +14,27 @@
-    "whitelist": [
-      "0010,0030"
-    ]
+  "DataAnon": {
+    "HardlinksUseHashBins" : true,
+    "Hardlinks": {
+      "/by-study-date/": "0008,0020",
+      "/by-pid/": "0010,0020",
+      "/by-dob/": "0010,0030"
+    },
+    "DateTruncation": {
+      "default": "YYYYMM01",
+      "0010,0030": "YYY0MM01"
+    },
+    "Filter": {
+      "blacklist": [
+        "0010"
+      ],
+      "whitelist": [
+        "0010,0030"
+      ]
+    }
   },
```
This would have been nice to move into its own config file. Then we could have detected it's presence and created a default file if missing.

cppcooper commented on include/dicom/dicom-element-view.h:
```diff
@@ -18,33 +18,37 @@
-    const uint32_t &tag = *(uint32_t *) (buffer + idx);
-    const uint16_t &group = *(uint16_t *) (buffer + idx);
-    const uint16_t &element = *(uint16_t *) (buffer + idx + 2);;
+    // these are apparently UB, according to the experts on TPH (discord) (ie. `*(uint32_t*) (buffer + idx);`)
+    const uint32_t tag = ReadTag();
+    const uint16_t group = tag&GROUP_MASK;
+    const uint16_t element = tag-group;
```
darn memory alignment and UB

cppcooper commented on include/dicom/dicom-tag.h:
```diff
@@ -0,0 +1,27 @@
+
+// Dicom Tags/VR/Name found here https://www.dicomlibrary.com/dicom/dicom-tags/
+enum DicomTag{
+    StudyDate = 0x00200008,
+    PName = 0x00100010,
+    PID = 0x00200010,
+    DOB = 0x00300010,
+    Item = 0xe000fffe,
+    ItemDelItem = 0xe00dfffe,
+    SeqDelItem = 0xe0ddfffe
```
This could in theory be modified to read in `group,element` order, we'd just have to do stuff to the enums where we use them. It wouldn't be good, but we could do it

cppcooper commented on include/plugin/data-transport.h:
```diff
@@ -0,0 +1,30 @@
+    static std::unordered_map<const void*, DicomFile> file_map;
+    static std::unordered_set<Md5> hashes_map;
+public:
+    static void Emplace(const void* instance_data, std::string md5, size_t size);
+    static void Emplace(const void* instance_data, std::string uuid);
+    static void Emplace(const void* instance_data, const DicomFile& file);
+    static bool Emplace(std::string md5);
```
These in particular really need better names.. I think. I think?

cppcooper commented on include/plugin/job-queue.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
+/*
+#include <mutex>
+#include <atomic>
+#include <queue>
+#include <condition_variable>
+#include <functional>
+
+class JobQueue {
+private:
+    std::atomic_bool keep_running;
+    std::atomic_bool has_work;
+    std::condition_variable cv;
+    std::mutex queue_lock;
+    std::queue<std::function<void()>> jqueue;
+    // this class is the only thing that can instantiate itself now
+    JobQueue(){}
+public:
+    static JobQueue& GetInstance();
+    void AddJob(std::function<void()> job);
+    void Process();
+    void Stop();
+};
+/**/
```
These two files should be deleted before merging to master

cppcooper replied:
actually 3 (.h .cpp, and the unit test .cpp)

cppcooper commented on src/plugin.cpp:
```diff
@@ -12,31 +15,46 @@ namespace globals {
-        //OrthancPluginRegisterIncomingDicomInstanceFilter(context, FilterCallback);
+        OrthancPluginRegisterOnStoredInstanceCallback(context, OnStoredInstanceCallback);
+        OrthancPluginRegisterIncomingDicomInstanceFilter(context, FilterCallback);
+        db_init_job = std::thread([](){
+            std::this_thread::sleep_for(std::chrono::seconds(2));
+            if(!DBInterface::Initialize()){
+                std::cerr << "DB Initialize job failed. Terminating.." << std::endl;
+                std::abort(); // abnormal termination
+            }
```
This should be documented. It's there because Orthanc's tables haven't yet initialized (potentially) yet and so we can't initialize ours yet. But we need it to initialize **from** our plugin initialize function, so we spawn a thread with a delay. A comment should explain this

cppcooper commented on src/plugin/initialize.sql:
```diff
@@ -0,0 +1,241 @@
+CREATE TABLE IF NOT EXISTS crosswalk (
```
In an ideal world, we break this file up into smaller portions and load those files instead of copy pasting the code in this file as string in numerous locations.

cppcooper commented on tests/common.h:
```diff
@@ -1,7 +1,11 @@
 extern fs::path GetProjRoot();
 extern void TestWithDicomFiles(std::function<void(const fs::path&)> test);
 namespace uuid {
     extern std::string generate_uuid_v4();
```
I don't think this function is used. It should be removed if that's the case.

cppcooper reviewed (changes requested):
It works, and is good.. but I think a few of these things should be addressed before merging to master. Most are fine as is.

brianzhouzc committed:
Update download-plugins.sh

cppcooper committed:
Merge pull request #120 from BCCF-UBCO-AD/hotfix_docker

Hotfix docker
